File: lib/models/notification_models.dart
import 'package:flutter/foundation.dart';

enum NotificationType { lowFood, clogged, cleaningDue, heavyUse, weatherBased, foodLevel }

enum UsageSensitivity { low, medium, high }

enum WeatherSensitivity { normal, high }

class NotificationEvent {
  final NotificationType type;
  final String message;
  final DateTime timestamp;
  final Map<String, dynamic>? meta;

  NotificationEvent({required this.type, required this.message, required this.timestamp, this.meta});
}

class NotificationPreferences with ChangeNotifier {
  bool lowFoodEnabled;
  bool cloggedEnabled;
  bool cleaningReminderEnabled;
  int cleaningIntervalDays;
  DateTime? lastCleaned;
  bool weatherBasedCleaningEnabled;
  WeatherSensitivity weatherSensitivity;
  double humidityThreshold;
  bool heavyUseEnabled;
  UsageSensitivity heavyUseSensitivity;
  double lowFoodThresholdPercent;
  bool progressNotificationsEnabled;
  double heavyUseCooldownHours;
  double weatherCooldownHours;

  NotificationPreferences({
    this.lowFoodEnabled = true,
    this.cloggedEnabled = true,
    this.cleaningReminderEnabled = true,
    this.cleaningIntervalDays = 7,
    this.lastCleaned,
    this.weatherBasedCleaningEnabled = true,
    this.weatherSensitivity = WeatherSensitivity.normal,
    this.humidityThreshold = 78,
    this.heavyUseEnabled = true,
    this.heavyUseSensitivity = UsageSensitivity.medium,
    this.lowFoodThresholdPercent = 20,
    this.progressNotificationsEnabled = true,
    this.heavyUseCooldownHours = 12,
    this.weatherCooldownHours = 12,
  });

  NotificationPreferences copyWith({
    bool? lowFoodEnabled,
    bool? cloggedEnabled,
    bool? cleaningReminderEnabled,
    int? cleaningIntervalDays,
    DateTime? lastCleaned,
    bool? weatherBasedCleaningEnabled,
    WeatherSensitivity? weatherSensitivity,
    double? humidityThreshold,
    bool? heavyUseEnabled,
    UsageSensitivity? heavyUseSensitivity,
    double? lowFoodThresholdPercent,
    bool? progressNotificationsEnabled,
    double? heavyUseCooldownHours,
    double? weatherCooldownHours,
  }) {
    return NotificationPreferences(
      lowFoodEnabled: lowFoodEnabled ?? this.lowFoodEnabled,
      cloggedEnabled: cloggedEnabled ?? this.cloggedEnabled,
      cleaningReminderEnabled: cleaningReminderEnabled ?? this.cleaningReminderEnabled,
      cleaningIntervalDays: cleaningIntervalDays ?? this.cleaningIntervalDays,
      lastCleaned: lastCleaned ?? this.lastCleaned,
      weatherBasedCleaningEnabled: weatherBasedCleaningEnabled ?? this.weatherBasedCleaningEnabled,
      weatherSensitivity: weatherSensitivity ?? this.weatherSensitivity,
      humidityThreshold: humidityThreshold ?? this.humidityThreshold,
      heavyUseEnabled: heavyUseEnabled ?? this.heavyUseEnabled,
      heavyUseSensitivity: heavyUseSensitivity ?? this.heavyUseSensitivity,
      lowFoodThresholdPercent: lowFoodThresholdPercent ?? this.lowFoodThresholdPercent,
      progressNotificationsEnabled: progressNotificationsEnabled ?? this.progressNotificationsEnabled,
      heavyUseCooldownHours: heavyUseCooldownHours ?? this.heavyUseCooldownHours,
      weatherCooldownHours: weatherCooldownHours ?? this.weatherCooldownHours,
    );
  }
}

File: ios/OrnimetricsNative/Sources/OrnimetricsNative/Models/NotificationModels.swift
import Foundation

enum NotificationType: String, Codable, CaseIterable, Identifiable {
    case lowFood
    case clogged
    case cleaningDue
    case weatherBased
    case heavyUse

    var id: String { rawValue }

    var title: String {
        switch self {
        case .lowFood: return "Low food"
        case .clogged: return "Clogged feeder"
        case .cleaningDue: return "Cleaning due"
        case .weatherBased: return "Weather cleaning"
        case .heavyUse: return "Heavy use"
        }
    }
}

enum WeatherSensitivity: String, Codable, CaseIterable, Identifiable {
    case normal
    case high

    var id: String { rawValue }
}

enum UsageSensitivity: String, Codable, CaseIterable, Identifiable {
    case low
    case normal
    case high

    var id: String { rawValue }
}

struct NotificationPreferences: Codable {
    var lowFoodEnabled: Bool = true
    var cloggedEnabled: Bool = true
    var cleaningReminderEnabled: Bool = true
    var cleaningIntervalDays: Int = 7
    var weatherBasedCleaningEnabled: Bool = true
    var weatherSensitivity: WeatherSensitivity = .normal
    var humidityThreshold: Double = 78
    var heavyUseEnabled: Bool = true
    var heavyUseSensitivity: UsageSensitivity = .normal
    var lowFoodThresholdPercent: Double = 20
    var progressNotificationsEnabled: Bool = true
    var heavyUseCooldownHours: Double = 12
    var weatherCooldownHours: Double = 12
    var lastCleaned: Date? = nil
}

struct NotificationEvent: Identifiable, Codable {
    let id: UUID
    let type: NotificationType
    let message: String
    let timestamp: Date
    let meta: [String: String]

    init(type: NotificationType, message: String, timestamp: Date = Date(), meta: [String: String] = [:]) {
        self.id = UUID()
        self.type = type
        self.message = message
        self.timestamp = timestamp
        self.meta = meta
    }
}

struct FoodLevelReading: Identifiable {
    let id = UUID()
    let percentFull: Double
    let timestamp: Date
}
